### ISSUE TITLE
GRUCell hidden state bias

### ISSUE THREAD

**Issue by cisimon7** [patrick-kidger/equinox#1003]
Hello,

I noticed a difference between the PyTorch GRUCell and the equinox GRUCell in terms of the bias parameter for the hidden state. It seems the PyTorch cell has more parameters than the equinox cell

```python
torch_cell = torch.nn.GRUCell(5, 3)
jax_cell = eqx.nn.GRUCell(5, 3, key=key)
```
```python
for name, param in torch_cell.named_parameters():
    print(name, param.shape)
```

```
weight_ih torch.Size([9, 5])
weight_hh torch.Size([9, 3])
bias_ih torch.Size([9])
bias_hh torch.Size([9])
```

```python
print(
    jax.tree.map(
        lambda x: x.shape,
        eqx.filter(jax_cell, eqx.is_array)
    )
)
```

```
GRUCell(
  weight_ih=(9, 5),
  weight_hh=(9, 3),
  bias=(9,),
  bias_n=(3,),
  input_size=5,
  hidden_size=3,
  use_bias=True
)
```

I expected the bias_hh (for the PyTorch) and bias_n (in equinox) to have the same shape. Looking into the code I see where the difference occur:

```python
class GRUCell(Module, strict=True):
    ...
    def __call__(
        self, input: Array, hidden: Array, *, key: Optional[PRNGKeyArray] = None
    ):
        ...
        igates = jnp.split(self.weight_ih @ input + bias, 3)
        hgates = jnp.split(self.weight_hh @ hidden, 3)
        reset = jnn.sigmoid(igates[0] + hgates[0])
        inp = jnn.sigmoid(igates[1] + hgates[1])
        new = jnn.tanh(igates[2] + reset * (hgates[2] + bias_n))
        return new + inp * (hidden - new)
```

My guess is the PyTorch equivalence in Equinox should have been:
```python
class GRUCell(Module, strict=True):
    ...
    def __call__(
        self, input: Array, hidden: Array, *, key: Optional[PRNGKeyArray] = None
    ):
        ...
        igates = jnp.split(self.weight_ih @ input + bias, 3)
        hgates = jnp.split(self.weight_hh @ hidden + bias_n, 3)
        reset = jnn.sigmoid(igates[0] + hgates[0])
        inp = jnn.sigmoid(igates[1] + hgates[1])
        new = jnn.tanh(igates[2] + reset * hgates[2])
        return new + inp * (hidden - new)
```

with the shape of bias_n changed according. Won't this difference have an effect in performance between a PyTorch model and an Equinox model?

**Comment by patrick-kidger** [comment]
This is a known (minor) issue with PyTorch here. There's no need to have two biases, they both affect the same hidden state. It's a holdover from a mistake made in IIRC cuDNN or similar. (I think the main effect is that you functionally end up doubling the variance of your bias at initialisation, and double its learning rate, since you have two parameters doing the same thing.)
